src/condor_utils/file_transfer.cpp

```diff
@@ -2251,6 +2250,7 @@ FileTransfer::DoDownload( filesize_t *total_bytes_ptr, ReliSock *s)
 	// Start the main download loop. Read reply codes + filenames off a
 	// socket wire, s, then handle downloads according to the reply code.
 	for (;;) {
+		int rc = 0;
```

Comment:
It looks like setting `rc` to `0` at the top of loop breaks a bunch of assumptions implicit in `if` statements throughout the loop.  It seems like it'd be a smaller and safer change to instead directly address the problem, which is that we attempt to do the first URL file-transfer even if a previous (CEDAR) file-transfer has failed.  Rather than make this change, it seems like it would make more sense to change the condition on single-file URL transfers to check if `rc` is nonzero.

Reply:
> It looks like setting rc to 0 at the top of loop breaks a bunch of assumptions implicit in if statements throughout the loop

Can you point out a case where you're concerned about?

Here's my logic:
- I went through all the conditionals and didn't see any other case where `rc` depends on the value of the previous loop iteration other than the bug under discussion.
- I can't imagine where such a dependency exists and it's the correct logic.
- By declaring inside the loop and not initializing, I got the compiler to generate a warning compiler if there's a case where `rc` could be read when generating the error message but there's a code path where `rc` is uninitialized.  When I did this, only the bug under discussion here is flagged.

So, between a human review of the code and the compiler's static analysis, I feel comfortable with the change.

One possible improvement for this patch is to do the test I tried -- declare but not initialize `rc` inside the loop.  Then, explicitly set `rc` inside the file transfer case to avoid the current bug.  This way, if a future code refactor introduces another uninitialized read in the conditional, a compiler warning is generated.

Reply:
My understanding was that we continue the main file-transfer loop past failures because the file-transfer protocol did not allow the starter to interrupt the shadow when the shadow during input transfer.  I apparently got confused along the way some where, and a re-inspection shows that we don't skip subsequent CEDAR transfers because of previous failures; indeed, we probably can't, because we need read all the content bits off the wire before reading the next command anyway.

However, it looks to me like the intent of the code, or at least one side-effect, is to avoid executing plug-ins unnecessarily -- the starter has to read the URL off the wire, but it doesn't have to start a plug-in -- and resetting `rc` to `0` at the top of every loop prevents that.

I think it would make sense to move the declaration of `rc` into the loop, but not initialize it, and change the current `rc != GET_FILE_PLUGIN_FAILED` conditions to `rc != 0`.